walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -117,4 +95,66 @@
     return timeRanges;
   }
 
+  /**
+   * Returns a list containing all the TimeRanges that the given collection of attendees are available for given
+   * a collection of their events for that day.
+   */
+  private List<TimeRange> determineAvailableTime(Collection<Event> events, Collection<String> attendees, long meetingDuration) {
+    List<TimeRange> availableTime = new ArrayList<>();
+    Iterator<TimeRange> unavailableTimeIterator = determineUnavailableTime(events, attendees).iterator();
+    TimeRange currentEvent = (unavailableTimeIterator.hasNext()) ?  unavailableTimeIterator.next() : null;
+    if (currentEvent != null) {
+      if (TimeRange.START_OF_DAY < currentEvent.start()
+          && currentEvent.start() - TimeRange.START_OF_DAY >= meetingDuration) {
+        availableTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currentEvent.start(), false));
+      }
+
+      while (unavailableTimeIterator.hasNext()) {
+        TimeRange nextEvent = unavailableTimeIterator.next();
+
+        if (nextEvent.start() - currentEvent.end() >= meetingDuration) {
+          availableTime.add(TimeRange.fromStartEnd(currentEvent.end(), nextEvent.start(), false));
+        }
+        currentEvent = nextEvent;
+      }
+
+      if (currentEvent.end() < TimeRange.END_OF_DAY
+          && TimeRange.END_OF_DAY - currentEvent.end() + 1 >= meetingDuration) {
+        availableTime.add(TimeRange.fromStartEnd(currentEvent.end(), TimeRange.END_OF_DAY, true));
+      }
+    } else {  // There are no times taken in the day.
+      availableTime.add(TimeRange.WHOLE_DAY);
+    }
+    
+    return availableTime;
+  }
+
+  /**
+   * Given the set of meeting times, this method refits the potential meeting times to fit
+   * optional attendees with the schedule, if applicable.
+   */
+  private Collection<TimeRange> considerOptionals(
+      List<TimeRange> meetingTimes, Collection<Event> events, MeetingRequest request) {
+    if (request.getOptionalAttendees().isEmpty()) {
+      return meetingTimes;
+    }
+
+    Collection<String> allAttendees = new HashSet<>();;
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+
+    Collection<TimeRange> totalAvailableTime = 
+        determineAvailableTime(events, allAttendees, request.getDuration());
+    Iterator<TimeRange> totalAvailableTimeIterator = totalAvailableTime.iterator();
+    
+    if (!totalAvailableTimeIterator.hasNext()) {
```

Comment:
Please reverse the 'if' and 'else' so you can have a positive condition (I can see why you might want this order because this statement is basically `if (iterator.isEmpty())` but since that function doesn't exist for iterators and you have to use `if (!iterator.hasNext())` instead I would reverse the order.

```
if (totalAvailableTimeIterator.hasNext()) {
  ...
}
```